md5.c

```diff
@@ -137,301 +145,245 @@ Final(uint8_t len[4])
+	MD5F3(&a, &b, &c, &d, in + ( 1*4), &F3CONST[ 4*4],  4);
+	MD5F3(&d, &a, &b, &c, in + ( 4*4), &F3CONST[ 5*4], 11);
+	MD5F3(&c, &d, &a, &b, in + ( 7*4), &F3CONST[ 6*4], 16);
+	MD5F3(&b, &c, &d, &a, in + (10*4), &F3CONST[ 7*4], 23);
+	MD5F3(&a, &b, &c, &d, in + (13*4), &F3CONST[ 8*4],  4);
+	MD5F3(&d, &a, &b, &c, in + ( 0*4), &F3CONST[ 9*4], 11);
+	MD5F3(&c, &d, &a, &b, in + ( 3*4), &F3CONST[10*4], 16);
+	MD5F3(&b, &c, &d, &a, in + ( 6*4), &F3CONST[11*4], 23);
+	MD5F3(&a, &b, &c, &d, in + ( 9*4), &F3CONST[12*4],  4);
+	MD5F3(&d, &a, &b, &c, in + (12*4), &F3CONST[13*4], 11);
+	MD5F3(&c, &d, &a, &b, in + (15*4), &F3CONST[14*4], 16);
+	MD5F3(&b, &c, &d, &a, in + ( 2*4), &F3CONST[15*4], 23);
+
+	MD5F4(&a, &b, &c, &d, in + ( 0*4), &F4CONST[ 0*4],  6);
+	MD5F4(&d, &a, &b, &c, in + ( 7*4), &F4CONST[ 1*4], 10);
+	MD5F4(&c, &d, &a, &b, in + (14*4), &F4CONST[ 2*4], 15);
+	MD5F4(&b, &c, &d, &a, in + ( 5*4), &F4CONST[ 3*4], 21);
+	MD5F4(&a, &b, &c, &d, in + (12*4), &F4CONST[ 4*4],  6);
+	MD5F4(&d, &a, &b, &c, in + ( 3*4), &F4CONST[ 5*4], 10);
+	MD5F4(&c, &d, &a, &b, in + (10*4), &F4CONST[ 6*4], 15);
+	MD5F4(&b, &c, &d, &a, in + ( 1*4), &F4CONST[ 7*4], 21);
+	MD5F4(&a, &b, &c, &d, in + ( 8*4), &F4CONST[ 8*4],  6);
+	MD5F4(&d, &a, &b, &c, in + (15*4), &F4CONST[ 9*4], 10);
+	MD5F4(&c, &d, &a, &b, in + ( 6*4), &F4CONST[10*4], 15);
+	MD5F4(&b, &c, &d, &a, in + (13*4), &F4CONST[11*4], 21);
+	MD5F4(&a, &b, &c, &d, in + ( 4*4), &F4CONST[12*4],  6);
+	MD5F4(&d, &a, &b, &c, in + (11*4), &F4CONST[13*4], 10);
+	MD5F4(&c, &d, &a, &b, in + ( 2*4), &F4CONST[14*4], 15);
+	MD5F4(&b, &c, &d, &a, in + ( 9*4), &F4CONST[15*4], 21);
+
+	add32(&digest+ 0, &a);
+	add32(&digest+ 4, &b);
+	add32(&digest+ 8, &c);
+	add32(&digest+12, &d);
 }
 
 
 /* Simple test program. */
-#include <stdio.h>
-#include <stdlib.h>
-#include <errno.h>
-
-int
-main (int argc, char **argv)
 {
-	int j;
-
-	if (argc < 2)
-	{
-		fprintf (stderr, "usage: %s files-to-hash\n", argv[0]);
-		exit (1);
-	}
-
-	for (j = 1; j < argc; ++j)
-	{
-		int err;
-		size_t ret;
-		int i;
-
-		FILE *fp;
-		uint32_t l;
-		uint8_t len[4];
-		uint8_t *in;
-
-		fp = fopen(argv[j], "rb");
-		err = errno;
-		if (fp == NULL) {
-			fprintf(stderr, "%s\n", strerror(err));
-			exit(1);
-		}
-		l = 0;
-		if (fseek(fp, 0, SEEK_END) == 0) {
-			l = ftell(fp);
-			fseek(fp, 0, SEEK_SET);
-		}
-
-		in = malloc(l);
-		err = errno;
-		if (in == NULL) {
-			fprintf(stderr, "%s\n", strerror(err));
-			exit(1);
-		}
-		ret = fread(in, 1, l, fp);
-		err = errno;
-		if (ret != l) {
-			fprintf(stderr, "%s\n", strerror(err));
-			exit(1);
-		}
-		fclose(fp);
-
-		len[0] = l; len[1] = l >> 8; len[2] = l >> 16; len[3] = l >> 24;
-		Init();
-		Update(in, l);
-		Final(len);
-		free(in);
-
-		for (i = 0; i < 16; i++)
-		{
-			printf("%02x", digest[i]);
-		}
-		printf("  %s\n", argv[j]);
-	}
-
-	return 0;
+	/* S3.ROM: 256KB = 0xC0000 - 0xFFFFF */
+	len20 = 0x04000;
+	len[0] = 0; len[1] = 0x40; len[2] = 0; len[3] = 0;

```

Comment:
Now only 16KB processed.
It takes 2[min] at x10 mode 😢 (2 * 10 * 16 = 320[min] = 5h20m on the device!)
How to get up speed?

Reply:
optimize with "[];" :

16KB: 33[sec]